spyder_unittest/backend/testrunner.py

```diff
@@ -0,0 +1,206 @@
+        Process running the unit test suite.
+    resultfilename : str
+        Name of file in which test results are stored.
+
+    Signals
+    -------
+    sig_finished(list of TestResult, str)
+        Emitted when test process finishes. First argument contains the test
+        results, second argument contains the output of the test process.
+    """
+
+    sig_finished = Signal(object, str)
+
+    def __init__(self, widget, resultfilename=None):
+        """
+        Construct test runner.
+
+        Parameters
+        ----------
+        widget : UnitTestWidget
+            Unit test widget which constructs the test runner.
+        resultfilename : str or None
+            Name of file in which to store test results. If None, use default.
+        """
+
+        QObject.__init__(self, widget)
+        self.process = None
+        if resultfilename is None:
+            self.resultfilename = os.path.join(tempfile.gettempdir(),
+                                               'unittest.results')
+        else:
+            self.resultfilename = resultfilename
+
+    def start(self, config, pythonpath):
+        """
+        Start process which will run the unit test suite.
+
+        The process is run in the working directory specified in 'config',
+        with the directories in `pythonpath` added to the Python path for the
+        test process. The test results are written to the file
+        `self.resultfilename`. The standard output and error are also recorded.
+        Once the process is finished, `self.finished()` will be called.
+
+        Parameters
+        ----------
+        config : TestConfig
+            Unit test configuration.
+        pythonpath : list of str
+            List of directories to be added to the Python path
+
+        Raises
+        ------
+        RuntimeError
+            If process failed to start.
+        """
+
+        framework = config.framework
+        wdir = config.wdir
+
+        self.process = QProcess(self)
+        self.process.setProcessChannelMode(QProcess.MergedChannels)
+        self.process.setWorkingDirectory(wdir)
+        self.process.finished.connect(self.finished)
+
+        if pythonpath is not None:
+            env = [
+                to_text_string(_pth)
+                for _pth in self.process.systemEnvironment()
+            ]
+            add_pathlist_to_PYTHONPATH(env, pythonpath)
+            processEnvironment = QProcessEnvironment()
+            for envItem in env:
+                envName, separator, envValue = envItem.partition('=')
+                processEnvironment.insert(envName, envValue)
+            self.process.setProcessEnvironment(processEnvironment)
+
+        if framework == 'nose':
+            executable = "nosetests"
+            p_args = ['--with-xunit', "--xunit-file=%s" % self.resultfilename]
+        elif framework == 'py.test':
+            executable = "py.test"
+            p_args = ['--junit-xml', self.resultfilename]
+        else:
```

Comment:
Another option to make this more extendable is to define subclasses that define some class variables

To see what I mean take a look at [this](https://github.com/ContinuumIO/ciocheck/blob/master/ciocheck/linters.py)